storagenode/storagenodedb/pieceinfo.go

```diff
@@ -59,41 +54,61 @@ func (db *pieceinfo) Add(ctx context.Context, info *pieces.Info) (err error) {
 		VALUES (?,?,?,?,?,?,?,?)
 	`), info.SatelliteID, info.PieceID, info.PieceSize, info.PieceCreation.UTC(), pieceExpiration, orderLimit, uplinkPieceHash, 0)
 
-	if err == nil {
-		db.loadSpaceUsed(ctx)
-		atomic.AddInt64(&db.usedSpace, info.PieceSize)
-	}
 	return ErrInfo.Wrap(err)
 }
 
-// GetPieceIDs gets pieceIDs using the satelliteID
-func (db *pieceinfo) GetPieceIDs(ctx context.Context, satelliteID storj.NodeID, createdBefore time.Time, limit, offset int) (pieceIDs []storj.PieceID, err error) {
-	defer mon.Task()(&ctx)(&err)
-
+func (db *v0PieceInfo) getAllPiecesOwnedBy(ctx context.Context, blobStore storage.Blobs, satelliteID storj.NodeID) ([]v0StoredPieceAccess, error) {
 	rows, err := db.db.QueryContext(ctx, db.Rebind(`
-		SELECT piece_id
+		SELECT piece_id, piece_size, piece_creation, piece_expiration
 		FROM pieceinfo_
-		WHERE satellite_id = ? AND datetime(piece_creation) < datetime(?)
+		WHERE satellite_id = ?
 		ORDER BY piece_id
-		LIMIT ? OFFSET ?
-	`), satelliteID, createdBefore.UTC(), limit, offset)
+	`), satelliteID)
 	if err != nil {
 		return nil, ErrInfo.Wrap(err)
 	}
 	defer func() { err = errs.Combine(err, rows.Close()) }()
+	var pieceInfos []v0StoredPieceAccess
 	for rows.Next() {
-		var pieceID storj.PieceID
-		err = rows.Scan(&pieceID)
+		pieceInfos = append(pieceInfos, v0StoredPieceAccess{
+			blobStore: blobStore,
+			satellite: satelliteID,
+		})
+		thisAccess := &pieceInfos[len(pieceInfos)-1]
+		err = rows.Scan(&thisAccess.pieceID, &thisAccess.pieceSize, &thisAccess.creationTime, &thisAccess.expirationTime)
 		if err != nil {
-			return pieceIDs, ErrInfo.Wrap(err)
+			return nil, ErrInfo.Wrap(err)
+		}
+	}
+	return pieceInfos, nil
+}
+
+// ForAllV0PieceIDsOwnedBySatellite executes doForEach for each locally stored piece, stored with
+// storage format V0 in the namespace of the given satellite. If doForEach returns a non-nil error,
+// ForAllV0PieceIDsOwnedBySatellite will stop iterating and return the error immediately.
+//
+// If blobStore is nil, the .Stat() and .FullPath() methods of the provided StoredPieceAccess
+// instances will not work, but otherwise everything should be ok.
+func (db *v0PieceInfo) ForAllV0PieceIDsOwnedBySatellite(ctx context.Context, blobStore storage.Blobs, satelliteID storj.NodeID, doForEach func(pieces.StoredPieceAccess) error) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	// TODO: is it worth paging this query? we hope that SNs will not yet have too many V0 pieces.
+	pieceInfos, err := db.getAllPiecesOwnedBy(ctx, blobStore, satelliteID)
+	if err != nil {
+		return err
+	}
+	// note we must not keep a transaction open with the db when calling doForEach; the callback
+	// might need to make db calls as well
+	for i := range pieceInfos {
+		if err := doForEach(&pieceInfos[i]); err != nil {
```

Comment:
Same question I asked above, is it worth adding any concurrency here for faster performance?

Reply:
Unlikely. Where needed, a doForEach callback can launch goroutines as desired. I don't expect we'll ever need the minimal amount of extra performance we'd be able to squeeze out of parallelizing the directory traversals themselves.

Reply:
:+1: